File: javascript/misc/diagonal-traverse.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var findDiagonalOrder = function(matrix) {
  up = true
  y = 0
  x = 0
  result = []
  while (y < matrix.length && x < matrix[0].length) {
    result.push(matrix[y][x])
    if (up) {
      if (x + 1 < matrix[0].length && y - 1 >= 0) {
        x++
        y--
      } else if (x + 1 >= matrix[0].length) {
        up = false
        y++
      } else if (y - 1 < 0) {
        up = false
        x++
      }
    } else {
      if (y + 1 < matrix.length && x - 1 >= 0) {
        x--
        y++
      } else if (y + 1 >= matrix.length) {
        up = true
        x++
      } else if (x - 1 < 0) {
        up = true
        y++
      }
    }
  }
  return result
};

File: ruby/leetcode/diagonal-traverse.rb
# @param {Integer[][]} matrix
# @return {Integer[]}
def find_diagonal_order(matrix)
    up = true # current direction of traversal
    y = 0
    x = 0
    result = []
    while y < matrix.size() && x < matrix[0].size()
        result << matrix[y][x]
        if up
            if x + 1 < matrix[0].size() && y - 1 >= 0
                x = x + 1
                y = y - 1
            elsif x + 1 >= matrix[0].size()
                up = false
                y = y + 1
            elsif y - 1 < 0
                up = false
                x = x + 1
            end
        elsif !up
            if y + 1 < matrix.size() && x - 1 >= 0
                x = x - 1
                y = y + 1
            elsif y + 1 >= matrix.size()
                up = true
                x = x + 1
            elsif x - 1 < 0
                up = true
                y = y + 1
            end
        end
    end
    return result
end
